File: tests/src/test/java/com/optivem/atddaccelerator/templategenerator/systemtest/clients/GithubClient.java
package com.optivem.atddaccelerator.templategenerator.systemtest.clients;

import com.optivem.atddaccelerator.templategenerator.systemtest.util.process.ProcessResult;
import org.springframework.http.HttpStatus;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import static com.optivem.atddaccelerator.templategenerator.systemtest.util.process.ProcessExecutor.*;
import static com.optivem.atddaccelerator.templategenerator.systemtest.util.process.ProcessResultAssertions.assertFailure;
import static com.optivem.atddaccelerator.templategenerator.systemtest.util.process.ProcessResultAssertions.assertSuccess;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class GithubClient {

    private String owner;
    private String repoName;
    private String repositoryPath;

    public GithubClient(String owner, String repoName) {
        this.owner = owner;
        this.repoName = repoName;
        this.repositoryPath = owner + "/" + repoName;
    }

    public String getRepositoryPath() {
        return repositoryPath;
    }

    public ProcessResult deleteRepository() {
        return executeProcess("gh", "repo", "delete", repositoryPath, "--yes");
    }

    public ProcessResult viewRepository() {
        return executeProcess("gh", "repo", "view", repositoryPath);
    }

    public ProcessResult viewPath(String path) {
        return executeProcess("gh", "api", "/repos/" + repositoryPath + "/contents/" + path);
    }

    public ProcessResult viewPages() {
        return executeProcess("gh", "api", "/repos/" + repositoryPath + "/pages");
    }

    public String getFileContent(String filePath) {
        var result = executeProcess("gh", "api",
                "-H", "Accept: application/vnd.github.raw",
                "/repos/" + repositoryPath + "/contents/" + filePath);
        assertSuccess(result);
        return result.getOutput();
    }

    public ProcessResult viewWorkflowRuns(String workflowFileName) {
        return executeProcess("gh", "run", "list",
                "--workflow", workflowFileName,
                "--limit", "1",
                "--json", "status,conclusion",
                "--repo", repositoryPath);
    }

    public boolean packageExists(String packageName) {
        var url = String.format("https://github.com/%s/pkgs/container/%s%2F%s", repositoryPath, repoName, packageName);
        return urlExists(url);
    }

    private boolean urlExists(String url) {
        WebClient client = WebClient.create();
        try {
            int statusCode = client.head()
                    .uri(url)
                    .exchangeToMono(response -> Mono.just(response.statusCode().value()))
                    .block();
            return statusCode == 200;
        } catch (Exception e) {
            return false;
        }
    }
}

File: console/tests/Clients/GithubClient.cs
using System.Net.Http;
using Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Util.Process;
using static Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Util.Process.ProcessExecutor;
using static Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Util.Process.ProcessResultAssertions;

namespace Optivem.AtddAccelerator.TemplateGenerator.SystemTests.Clients
{
    public class GithubClient
    {
        private readonly string _owner;
        private readonly string _repoName;
        private readonly string _repositoryPath;

        public GithubClient(string owner, string repoName)
        {
            _owner = owner;
            _repoName = repoName;
            _repositoryPath = $"{owner}/{repoName}";
        }

        public string GetRepositoryPath() => _repositoryPath;

        public ProcessResult DeleteRepository()
        {
            return ExecuteProcess("gh", "repo", "delete", _repositoryPath, "--yes");
        }

        public ProcessResult ViewRepository()
        {
            return ExecuteProcess("gh", "repo", "view", _repositoryPath);
        }

        public ProcessResult ViewPath(string path)
        {
            return ExecuteProcess("gh", "api", $"/repos/{_repositoryPath}/contents/{path}");
        }

        public ProcessResult ViewPages()
        {
            return ExecuteProcess("gh", "api", $"/repos/{_repositoryPath}/pages");
        }

        public string GetFileContent(string filePath)
        {
            var result = ExecuteProcess("gh", "api",
                "-H", "Accept: application/vnd.github.raw",
                $"/repos/{_repositoryPath}/contents/{filePath}");
            AssertSuccess(result);
            return result.Output;
        }

        public ProcessResult ViewWorkflowRuns(string workflowFileName)
        {
            return ExecuteProcess("gh", "run", "list",
                "--workflow", workflowFileName,
                "--limit", "1",
                "--json", "status,conclusion",
                "--repo", _repositoryPath);
        }

        public bool PackageExists(string packageName)
        {
            var url = $"https://github.com/{_repositoryPath}/pkgs/container/{_repoName}/{packageName}";
            return UrlExists(url);
        }

        private bool UrlExists(string url)
        {
            using var client = new HttpClient();
            try
            {
                var response = client.SendAsync(new HttpRequestMessage(HttpMethod.Head, url)).Result;
                return response.IsSuccessStatusCode;
            }
            catch (Exception)
            {
                return false;
            }
        }
    }
}